File: 2022/ts/day11.ts
import { assert } from "./deps.ts";
import { lcm, readLines } from "./utils.ts";

interface Monkey {
  items: number[];
  op: (old: number) => number;
  div: number;
  testTrue: number;
  testFalse: number;
  handled: number;
}

function toMonkey(monkeyData: string): Monkey {
  const lines = monkeyData.split("\n");
  const items = lines[1].split(": ")[1].split(", ").map((t) => parseInt(t));
  const op = (old: number) => {
    return eval(lines[2].split(" = ")[1].replaceAll("old", `${old}`));
  };
  const div = parseInt(lines[3].split(" by ")[1]);
  const testTrue = parseInt(lines[4].split(" monkey ")[1]);
  const testFalse = parseInt(lines[5].split(" monkey ")[1]);
  const handled = 0;
  return { items, op, div, testTrue, testFalse, handled };
}

function run(monkeys: Monkey[], steps: number, part2 = false): number {
  let modulus = monkeys[0].div;
  for (let i = 1; i < monkeys.length; ++i) {
    modulus = lcm(modulus, monkeys[i].div);
  }
  for (let i = 0; i < steps; ++i) {
    for (const monkey of monkeys) {
      for (const item of monkey.items) {
        let worryLevel = item;
        if (part2) {
          worryLevel = monkey.op(worryLevel) % modulus;
        } else {
          worryLevel = Math.floor(monkey.op(worryLevel) / 3);
        }
        if (worryLevel % monkey.div === 0) {
          monkeys[monkey.testTrue].items.push(worryLevel);
        } else {
          monkeys[monkey.testFalse].items.push(worryLevel);
        }
        monkey.handled++;
        monkey.items = [];
      }
    }
  }
  monkeys.sort((a, b) => b.handled - a.handled);
  return monkeys[0].handled * monkeys[1].handled;
}

function part1() {
  const monkeys = readLines("../inputs/input11", "\n\n").map((chunk) =>
    toMonkey(chunk)
  );
  const part1 = run(monkeys, 20);
  console.log(`Part 1: ${part1}`);
}

part1();

function part2() {
  const monkeys = readLines("../inputs/input11", "\n\n").map((chunk) =>
    toMonkey(chunk)
  );
  const part2 = run(monkeys, 10000, true);
  console.log(`Part 2: ${part2}`);
}

part2();

File: 2022/python/day11.py
from copy import deepcopy
from math import lcm

def to_op(input):
	return lambda x: eval(input, {}, {"old": x})

def run(monkeys, steps, part2=False):
	monkeys = deepcopy(monkeys)
	divs = []
	for i in monkeys:
		divs.append(monkeys[i][2])
	modulus = lcm(*divs)
	for _ in range(steps):
		for i in monkeys:
			monkey = monkeys[i]
			for item in monkey[0]:
				monkey[4] = monkey[4]+1
				if part2:
					worrylevel = monkey[1](item) % modulus
				else:
					worrylevel = monkey[1](item) // 3
				if worrylevel % monkey[2] == 0:
					monkeys[monkey[3][0]][0].append(worrylevel)
				else:
					monkeys[monkey[3][1]][0].append(worrylevel)

			monkey[0] = []

	top_two = sorted(monkeys.items(), key=lambda item: item[1][4], reverse=True)[0:2]
	print(top_two[0][1][4] * top_two[1][1][4])

with open('../inputs/input11') as f:
	monkey_dict = {}
	monkeys = f.read().split('\n\n')
	for monkey in monkeys:
		entries = monkey.split('\n')
		i = int(entries[0].split(' ')[1].replace(':', ''))
		items = [int(t) for t in entries[1].split(': ')[1].split(', ')]
		op = to_op(entries[2].split(' = ')[1])
		test = int(entries[3].split(' by ')[1])
		test_true = int(entries[4].split('monkey ')[1])
		test_false = int(entries[5].split('monkey ')[1])
		monkey_dict[i]  = [items, op, test, [test_true, test_false], 0]
	run(monkey_dict, 20)
	run(monkey_dict, 10000, True)